### ISSUE TITLE
Unstable training with MQA and TP>1

### ISSUE THREAD

**Issue by RaymondLi0** [bigcode-project/Megatron-LM#35]
Models trained with MQA and TP>1 have exploding loss after a few thousand steps.
This issue does not happen when training the same models with MHA, or with TP=1.

<img width="459" alt="image (1)" src="https://user-images.githubusercontent.com/11444023/226378518-4c3fd932-ffab-4715-9475-b28d8a5dff9b.png">

**Comment by RaymondLi0** [comment]
Green: TP=1
Purple: TP=2

At step 2400, the weights across TP-ranks are still perfectly in sync:
```
==> logs/logs_rank_0.txt <==
2023-03-20 21:50:27,016 [Rank 0]: Iteration 2400
2023-03-20 21:50:27,017 [Rank 0]: Rank 0, model 0, layer 0, weight -0.0001285076141357422, bias -0.0006017684936523438, weight max 0.13232421875
2023-03-20 21:50:27,017 [Rank 0]: Rank 0, model 0, layer 1, weight 0.00013685226440429688, bias -0.0006756782531738281, weight max 0.1378173828125
2023-03-20 21:50:27,018 [Rank 0]: Rank 0, model 0, layer 2, weight -3.1113624572753906e-05, bias -0.0008378028869628906, weight max 0.1600341796875
2023-03-20 21:50:27,018 [Rank 0]: Rank 0, model 0, layer 3, weight -1.0311603546142578e-05, bias 0.0004892349243164062, weight max 0.1666259765625
2023-03-20 21:50:27,018 [Rank 0]: Rank 0, model 0, layer 4, weight 0.00024068355560302734, bias 0.000469207763671875, weight max 0.1748046875
2023-03-20 21:50:27,019 [Rank 0]: Rank 0, model 0, layer 5, weight -6.896257400512695e-05, bias 0.003726959228515625, weight max 0.169921875
2023-03-20 21:50:27,019 [Rank 0]: Rank 0, model 0, layer 6, weight 6.204843521118164e-05, bias 0.008575439453125, weight max 0.170654296875
2023-03-20 21:50:27,019 [Rank 0]: Rank 0, model 0, layer 7, weight 7.802248001098633e-05, bias -0.00884246826171875, weight max 0.1578369140625

==> logs/logs_rank_1.txt <==
2023-03-20 21:50:27,016 [Rank 1]: Iteration 2400
2023-03-20 21:50:27,017 [Rank 1]: Rank 1, model 0, layer 0, weight -0.0001285076141357422, bias -0.0006017684936523438, weight max 0.13232421875
2023-03-20 21:50:27,017 [Rank 1]: Rank 1, model 0, layer 1, weight 0.00013685226440429688, bias -0.0006756782531738281, weight max 0.1378173828125
2023-03-20 21:50:27,017 [Rank 1]: Rank 1, model 0, layer 2, weight -3.1113624572753906e-05, bias -0.0008378028869628906, weight max 0.1600341796875
2023-03-20 21:50:27,018 [Rank 1]: Rank 1, model 0, layer 3, weight -1.0311603546142578e-05, bias 0.0004892349243164062, weight max 0.1666259765625
2023-03-20 21:50:27,018 [Rank 1]: Rank 1, model 0, layer 4, weight 0.00024068355560302734, bias 0.000469207763671875, weight max 0.1748046875
2023-03-20 21:50:27,019 [Rank 1]: Rank 1, model 0, layer 5, weight -6.896257400512695e-05, bias 0.003726959228515625, weight max 0.169921875
2023-03-20 21:50:27,019 [Rank 1]: Rank 1, model 0, layer 6, weight 6.204843521118164e-05, bias 0.008575439453125, weight max 0.170654296875
2023-03-20 21:50:27,019 [Rank 1]: Rank 1, model 0, layer 7, weight 7.802248001098633e-05, bias -0.00884246826171875, weight max 0.1578369140625

[Goes on the same for other ranks]
```

![Screen Shot 2023-03-20 at 6 37 46 PM](https://user-images.githubusercontent.com/11444023/226483294-db04ed4d-f565-404f-8476-681ed88b97fd.png)

**Comment by RaymondLi0** [comment]
The issue also occurs with full-recomputation of the activations, or no recomputation at all

![Screen Shot 2023-03-20 at 6 53 57 PM](https://user-images.githubusercontent.com/11444023/226483681-15bdd192-f2da-4302-b796-5c780c1b0de5.png)

**Comment by RaymondLi0** [comment]
- The fix proposed in #36  seems to work! With TP=2 it gives the black curve that we can't even see because it's behind the green curve (TP=1)

- However with sequence-parallel (blue curve), there a significant difference

- Putting the reduction outside of the `else` as suggested by @jlamypoirier  seems to work (red curve)
```
--- a/megatron/model/transformer.py
+++ b/megatron/model/transformer.py
@@ -614,6 +614,8 @@ class ParallelAttention(MegatronModule):
             if get_args().sequence_parallel:
                 # The linear layer doesn't gather the sequence-parallel.
                 kv_input = mpu.gather_from_sequence_parallel_region(kv_input, tensor_parallel_output_grad=False)
+            # else:
+            kv_input = mpu.copy_to_tensor_model_parallel_region(kv_input)
             # Attention heads [sq, b, h] --> [sq, b, (2 * hn)]
             mixed_kv_layer = self.key_value(kv_input)
```

![Screen Shot 2023-03-20 at 11 24 53 PM](https://user-images.githubusercontent.com/11444023/226510877-8d3afcaf-4140-41d5-8935-e34699dc34a7.png)

**Comment by thomasw21** [comment]
Probably fixed with #39
